src/targets/gpu/device/nonzero.cpp

```diff
@@ -0,0 +1,66 @@
+#include <migraphx/gpu/device/nonzero.hpp>
+#include <migraphx/gpu/device/float_equal.hpp>
+#include <migraphx/gpu/device/scan.hpp>
+#include <migraphx/gpu/device/fill.hpp>
+#include <migraphx/gpu/device/reduce_ops.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace gpu {
+namespace device {
+
+argument nonzero(hipStream_t stream,
+                 const argument& result,
+                 const argument& arg_idx,
+                 const argument& arg_data)
+{
+    auto elem_num = arg_data.get_shape().elements();
+    // check nonzero elements
+    arg_data.visit([&](auto input) {
+        const auto* in_ptr = device_cast(input.data());
+        auto* idx          = arg_idx.cast<int64_t>();
+        gs_launch(stream, elem_num)(
+            [=](auto i) __device__ { idx[i] = (float_equal(in_ptr[i], 0)) ? 0 : 1; });
+    });
+
+    // set all output values to 0
+    fill(stream, result, 0);
+
+    // call the prefix_sum function to do a prefix_sum to compute
+    // index in the output. Only 1 block can be used since we have
+    // only one prefix sum
+    const index_int block_size = 256;
+    arg_idx.visit([&](auto in_idx) {
+        auto* ptr = device_cast(in_idx.data());
+        gs_launch(stream, block_size, block_size)([=](auto, auto idx) __device__ {
+            block_scan<block_size>(idx,
+                                   sum{},
+                                   0,
+                                   elem_num,
+                                   [&](auto j) { return ptr[j]; },
+                                   [&](auto j, auto x) { ptr[j] = x; });
```

Comment:
I wonder if you could write the output here:

```cpp
[&](auto j, auto x) { 
    auto out_idx = x - 1;
    auto index   = si.multi(out_idx);
    for(std::size_t k = 0; k < index.size(); ++k)
    {
        out_ptr[k * elem_num + out_idx] = index[k];
    }
}
```

Reply:
fixed

Reply:
After additional test, the out_idx in the line auto index   = si.multi(out_idx); should be i, and not every thread can write to the output. Only thread i with idx[i] != idx[i - 1] can write to the output.

Reply:
I dont follow, can you explain some more?

Reply:
Why cant you use `auto index  = si.multi(j)`?

Reply:
A simple example:
we have input 2, 0, 3, 0, 1 (5 elements), then the output should be
out[0] = 0, out[1] = 2, out[2] = 4.

In our implementation, after the prefix_sum, we got:
1, 1, 2, 2, 3

This means different j's have the same "x", and the different j's are written to the same output location since we use "x" to indicate where to write the "j". 

The approach I used is for each "j" after the prefix_sum, we compare its "x" to that of "j - 1". Only if the x's are different, we write "j" to the output.

Reply:
We shouldn't need to check the adjacent element, we can just check `not float_equal(ptr[j], 0)`.

Reply:
Good idea, changed. Thanks!